Distributed Index Analysis

azat opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Distributed index analysis on multiple replicas. Beneficial for shared storage and huge amount of data in cluster.

This is the first implementation of distributed index analysis for shared storage. Implemented via separate function (`mergeTreeAnalyzeIndexes()`), during index analysis, if `distributed_index_analysis=1`, it will offload index analysis to other replicas based on consistent hashing of part names (_the cluster name from `cluster_for_parallel_replicas` is used, also some other settings like `max_parallel_replicas` is used_). In case of remote replica is unavailable (or in case of any other errors), the index analysis will be done on the initiator.

It is beneficial for clusters with huge amount of data, when even PK is very big. `use_const_adaptive_granularity=1` is highly recommended (to avoid using too much memory for marks).

And also it requires `use_primary_key_cache=1`, to unload irrelevant indexes.


<!-- ch-version-info:start -->
### Version info
- Merged into: `26.1.1.646`
<!-- ch-version-info:end -->

EmeraldShift commented:
Im curious, what kinds of optimizations will this make possible that aren't currently done? How does index analysis currently work for Distributed tables? Any chance this could help with https://github.com/ClickHouse/ClickHouse/issues/89769?

azat committed:
Distributed Index Analysis

azat committed:
Do not run distributed index analysis multiple times in case of PR

azat commented on src/Core/Settings.cpp:
```diff
@@ -6964,6 +6964,12 @@ Replace table function engines with their -Cluster alternatives
+    DECLARE(Bool, distributed_index_analysis, false, R"(
+Index analysis will be distributed across replicas. Requires cluster_for_parallel_replicas.
+)", EXPERIMENTAL) \
+    DECLARE(Bool, parallel_replicas_use_distributed_index_analysis_info, false, R"(
```
This will be removed

azat committed:
Remove parallel_replicas_use_distributed_index_analysis_info

Not only it does not improve anything, but it also makes parallel
replicas less effective, due to with it replicas are more limited to
what ranges they can handle.

azat committed:
Extend EXPLAIN indexes=1 for distributed_index_analysis

KochetovNicolai commented on src/Interpreters/ClusterProxy/distributedIndexAnalysis.cpp:
```diff
@@ -0,0 +1,356 @@
+            runner.enqueueAndKeepTrack([&, i, replica_address]()
+            {
+                LOG_TRACE(logger, "Resolving {} parts ({} marks, {} rows) from local replica {} (index {}): {}", replica_parts.size(), replicas_marks[i], replicas_rows[i], replica_address, i, replica_parts);
+                auto parts_ranges = local_index_analysis_callback(replica_parts);
```
Maybe we can run it in the main thread, as we do for `missing_parts` ?

azat committed:
Distributed Index Analysis: remove max_result_rows/max_result_rows

azat committed:
Distributed Index Analysis: Pushdown Support

azat committed:
Distributed Index Analysis: Support GLOBAL IN

azat committed:
tests: fix 03621_pr_distributed_index_analysis flakiness

One query may have both, QueryFinish and ExceptionWhileProcessing [1].

  [1]: https://pastila.nl/?002138a7/10b6057da2af2d9cbb28827cbddd0c6b#zR1Jr/+tcOU9mT3Y9Y339g==GCM

azat committed:
tests: fix other failures of 03621_pr_distributed_index_analysis

azat committed:
Use PullingPipelineExecutor and RemoteSource to make it cancellable

KochetovNicolai commented on src/Interpreters/ClusterProxy/distributedIndexAnalysis.cpp:
```diff
@@ -0,0 +1,350 @@
+
+DistributedIndexAnalysisPartsRanges distributedIndexAnalysisOnReplicas(
+    const StorageID & storage_id,
+    const SelectQueryInfo & query_info,
```
It appears that only `query_info.filter_actions_dag` is used. Maybe make iface cleaned and pass only it?

azat replied:
Make sense, done

KochetovNicolai commented on src/Interpreters/ClusterProxy/distributedIndexAnalysis.cpp:
```diff
@@ -0,0 +1,350 @@
+    res.resize(replicas);
+
+    ContextMutablePtr execution_context = Context::createCopy(context);
+    auto external_tables = execution_context->getExternalTables();
```
I think now we can send external tables which could be not needed for the predicate.
If so, can we filter out and send only needed ones?

azat replied:
I think it should work correctly, because `primary_key_columns_names_set` is passed to `tryBuildAdditionalFilterAST`

Let me check, indeed - https://pastila.nl/?000ac056/6008649427d36b24f1ef2a5b862b1459#Z0HfK3STy6r1jgfGXeKviw==GCM

As you can see first query has `Connection (127.0.0.2:9000): Sent data for 1 external tables` while the second does not

KochetovNicolai reviewed (commented):
Overall looks good.

azat committed:
tests: disable 03621_pr_distributed_index_analysis for replicated-database

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=86786&sha=27233b7e7855bbe02f02b51320ff1217584cd459&name_0=PR&name_1=Stateless%20tests%20%28amd_binary%2C%20old%20analyzer%2C%20s3%20storage%2C%20DatabaseReplicated%2C%20parallel%29

azat committed:
Distributed Index Analysis: do not send filter predicate if there is no any

azat committed:
Distributed Index Analysis: pass filter_actions_dag instead of SelectQueryInfo

azat committed:
Distributed Index Analysis: Cleaner LocalIndexAnalysisCallback

azat committed:
Introduce IndexAnalysisContext for filterPartsByPrimaryKeyAndSkipIndexes()

azat committed:
mergeTreeAnalyzeIndexes: fix "Column BLOB is not a contiguous block of memory"

azat committed:
More details in "Expected ASTSelectWithUnionQuery, ASTSelectQuery or ASTSubquery" error

azat committed:
Fix GLOBAL IN with distributed_index_analysis and parallel replicas

azat committed:
Split LocalIndexAnalysisCallback lambda into separate function

azat committed:
tests: makr 03621_pr_distributed_index_analysis as long

CI: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=86786&sha=b3a9951be86e45876027cbe04a2a8068cc0ee19a&name_0=PR&name_1=Stateless%20tests%20%28amd_asan%2C%20flaky%20check%29

clickgapai commented:
Hi @azat @KochetovNicolai — while reviewing this PR I found the following:

- **Bug (P3):** `DistributedIndexAnalysisThreadsScheduled` metric description has typo "idnex" instead of "index" — https://github.com/ClickHouse/ClickHouse/issues/103317

Happy to discuss — close anything that's wrong or already addressed.
